content/blog/2018/04/2018-04-24-configure-jenkins-pipeline-with-yaml-file.adoc

```diff
@@ -0,0 +1,380 @@
+
+Groovy’s flexible enough to allow this same declarative code to be
+created at runtime, making our dream of using a YAML to configure our
+job come true!
+
+==== Introducing Wolox-CI
+
+That’s how https://github.com/Wolox/wolox-ci[wolox-ci] was born- our
+shared library for Jenkins!
+
+With https://github.com/Wolox/wolox-ci[wolox-ci], our Jenkinsfile is now
+reduced to:
+
+....
+@Library('wolox-ci') _
+
+node {
+
+  checkout scm
+
+  woloxCi('.woloxci/config.yml');
+}
+....
+
+Now it simply checks out the code and then calls wolox-ci. The library
+reads yaml file like this one
+
+....
+config:
+  dockerfile: .woloxci/Dockerfile
+  project_name: some-project-name
+
+services:
+  - postgresql
+  - redis
+
+steps:
+  analysis:
+    - bundle exec rubocop -R app spec --format simple
+    - bundle exec rubycritic --path ./analysis --minimum-score 80 --no-browser
+  setup_db:
+    - bundle exec rails db:create
+    - bundle exec rails db:schema:load
+  test:
+    - bundle exec rspec
+  security:
+    - bundle exec brakeman --exit-on-error
+  audit:
+    - bundle audit check --update
+
+
+environment:
+  RAILS_ENV: test
+  GIT_COMMITTER_NAME: a
+  GIT_COMMITTER_EMAIL: b
+  LANG: C.UTF-8
+....
+
+and builds the Jenkinsfile to get your job running on the fly.
+
+The nice part about having a shared library is that we can extend and
+fix our library in a centralized way. Once we add new code, the library
+is automatically updated in Jenkins which will notify all of our jobs
+with the update.
+
+Since we have projects in different languages we use Docker to build the
+testing environment. WoloxCI assumes there is a Dockerfile to build and
+will run all the specified commands inside the container.
+
+===== Woloxci config.yml
+
+====== Config
+
+The first part of the config.yml file specifies some basic
+configuration: project’s name and Dockerfile location. The Dockerfile is
+used to build the image where the commands will be run.
+
+====== Services
+
+This section describes which services will be exposed to the container.
+Out of the box, WoloxCI has support for _postgresql_, _mssql_ and
+_redis_. You can also specify the docker image version you want! It is
+not hard to add a new service. You just need to add the corresponding
+file at
+
+https://github.com/Wolox/wolox-ci/tree/development/vars
+
+and modify how the services are parsed
+
+https://github.com/Wolox/wolox-ci/blob/development/src/com/wolox/parser/ConfigParser.groovy#L76
+
+====== Steps
+
+The listed commands in this section will run inside the Docker
+container. As a result, you’ll see each of the steps on the Jenkins UI.
+
+image:https://cdn-images-1.medium.com/max/2000/0*SlHf1JHAAvEvZQ74.[image]
+
+====== Environment
+
+If you need some environment variables during your build, you can
+specify them here. Whatever variable you set will be available inside
+the Docker container when your commands listed in the *steps* section
+described above.
+
+=== Wrapping up
```

Comment:
I noticed that https://github.com/Wolox/wolox-ci has no contributing guidelines. Maybe it could be something to add and reference in the post (just in case somebody wants to extend it)

Reply:
Done